Fix size of text drawing using wxGraphicsContext when printing

vadz opened the pull request:
When using both wxEnhMetaFileDC (used for print preview) and
wxPrinterDC, a global scale factor is applied to the graphics context in
wxGDIPlusPrintingContext ctor and this changes size of the text rendered
using size specified in point units.

To prevent this from happening, use pixel units for the font size when
creating it (because we can't know if the font is going to be used with
screen or printer contexts and, in fact, it can be used with both), but
scale them correctly to obtain fonts of the same size as if we used
points for the screen contexts.

This fixes a regression with text size when printing or previewing
introduced by d5020362ffb99608eacd2492d621cad20b1075a7

Closes [#18338](https://trac.wxwidgets.org/ticket/18338).

a-wi reviewed (commented):
This trick would work but maybe another trick is possible because actual font scale is stored in `wxGDIPlusContext::m_fontScaleRatio` and this value is e.g. used to do manual "rescaling" in `wxGDIPlusContext::GetTextExtent`.
So, maybe drawing the text could be done in `wxGDIPlusContext::DoDrawText` with something like this:
`m_context->PushState();
m_context->Scale(m_fontScaleRatio,m_fontScaleRatio);
m_context->DrawString(...);
m_context->PopState();
`

vadz commented:
Something like this seems to work roughly correctly, but less exactly than the version in this PR, i.e. the text height is not huge as it is now, but not the same as when drawn by `wxDC` neither:
```diff
diff --git a/src/msw/graphics.cpp b/src/msw/graphics.cpp
index ebecbeb72a..682d94c7f9 100644
--- a/src/msw/graphics.cpp
+++ b/src/msw/graphics.cpp
@@ -2094,15 +2094,20 @@ void wxGDIPlusContext::DoDrawText(const wxString& str,
     wxGDIPlusFontData * const
         fontData = (wxGDIPlusFontData *)m_font.GetRefData();
 
+    PushState();
+    Scale(1/m_fontScaleRatio, 1/m_fontScaleRatio);
+
     m_context->DrawString
                (
                     str.wc_str(*wxConvUI),  // string to draw, always Unicode
                     -1,                     // length: string is NUL-terminated
                     fontData->GetGDIPlusFont(),
-                    PointF(x, y),
+                    PointF(x*m_fontScaleRatio, y*m_fontScaleRatio),
                     GetDrawTextStringFormat(),
                     fontData->GetGDIPlusBrush()
                );
+
+    PopState();
 }
 
 void wxGDIPlusContext::GetTextExtent( const wxString &str, wxDouble *width, wxDouble *height,
@@ -2116,8 +2121,6 @@ void wxGDIPlusContext::GetTextExtent( const wxString &str, wxDouble *width, wxDo
 
     f->GetFamily(&ffamily) ;
 
-    REAL factorY = m_fontScaleRatio;
-
     // Notice that we must use the real font style or the results would be
     // incorrect for italic/bold fonts.
     const INT style = f->GetStyle();
@@ -2127,17 +2130,17 @@ void wxGDIPlusContext::GetTextExtent( const wxString &str, wxDouble *width, wxDo
     REAL rAscent = ffamily.GetCellAscent(style) * size / emHeight;
     REAL rHeight = ffamily.GetLineSpacing(style) * size / emHeight;
 
-    if ( height )
-        *height = rHeight * factorY;
     if ( descent )
-        *descent = rDescent * factorY;
+        *descent = rDescent / m_fontScaleRatio;
     if ( externalLeading )
-        *externalLeading = (rHeight - rAscent - rDescent) * factorY;
+        *externalLeading = (rHeight - rAscent - rDescent) / m_fontScaleRatio;
     // measuring empty strings is not guaranteed, so do it by hand
     if ( str.IsEmpty())
     {
         if ( width )
             *width = 0 ;
+        if ( height )
+            *height = rHeight / m_fontScaleRatio;
     }
     else
     {
@@ -2146,9 +2149,9 @@ void wxGDIPlusContext::GetTextExtent( const wxString &str, wxDouble *width, wxDo
         RectF bounds ;
         m_context->MeasureString((const wchar_t *) s , wcslen(s) , f, layoutRect, GetDrawTextStringFormat(), &bounds ) ;
         if ( width )
-            *width = bounds.Width;
+            *width = bounds.Width / m_fontScaleRatio;
         if ( height )
-            *height = bounds.Height;
+            *height = bounds.Height / m_fontScaleRatio;
     }
 }
 
}}}

I'm not sure why exactly is it so, i.e. I'd expect it to work exactly the same too, but it doesn't. Do you have any idea?

a-wi commented:
Indeed, rescaling drawn text this doesn't give satisfactory results. What about real font rescaling in printing context? This seems to work for me:
```diff
diff --git a/src/msw/graphics.cpp b/src/msw/graphics.cpp
--- a/src/msw/graphics.cpp
+++ b/src/msw/graphics.cpp
@@ -508,6 +508,11 @@ class wxGDIPlusPrintingContext : public wxGDIPlusContext
 public:
     wxGDIPlusPrintingContext( wxGraphicsRenderer* renderer, const wxDC& dc );
     virtual ~wxGDIPlusPrintingContext() { }
+
+    // creates a font
+    virtual wxGraphicsFont CreateFont(const wxFont &font, const wxColour &col) const wxOVERRIDE;
+    virtual wxGraphicsFont CreateFont(double sizeInPixels, const wxString& facename,
+                                      int flags, const wxColour& col) const wxOVERRIDE;
 protected:
 };
 
@@ -2116,7 +2121,7 @@ void wxGDIPlusContext::GetTextExtent( const wxString &str, wxDouble *width, wxDo
 
     f->GetFamily(&ffamily) ;
 
-    REAL factorY = m_fontScaleRatio;
+    REAL factorY = 1.0;// m_fontScaleRatio;
 
     // Notice that we must use the real font style or the results would be
     // incorrect for italic/bold fonts.
@@ -2271,7 +2276,20 @@ wxGDIPlusPrintingContext::wxGDIPlusPrintingContext( wxGraphicsRenderer* renderer
 
     // We use this modifier when measuring fonts. It is needed because the
     // page scale is modified above.
-    m_fontScaleRatio = context->GetDpiY() / 72.0;
+    m_fontScaleRatio = context->GetDpiY() / 96.0;
+}
+
+wxGraphicsFont wxGDIPlusPrintingContext::CreateFont(const wxFont &font, const wxColour &col) const
+{
+    wxFont f(font);
+    f.SetFractionalPointSize(font.GetFractionalPointSize() / m_fontScaleRatio);
+    return wxGDIPlusContext::CreateFont(f, col);
+}
+
+wxGraphicsFont wxGDIPlusPrintingContext::CreateFont(double size, const wxString& facename,
+                                                    int flags, const wxColour& col) const
+{
+    return wxGDIPlusContext::CreateFont(size / m_fontScaleRatio, facename, flags, col);
 }
 
 //-----------------------------------------------------------------------------
```

vadz commented:
I thought of this originally but discarded this idea because I thought that we couldn't rely on the font being created using this particular context. In the printing sample this would work because we just use a `wxFont` and so a new `wxGraphicsFont` is created every time, but in principle we could have created `wxGraphicsFont` once and then use it with different contexts, couldn't we? Or is a `wxGraphicsFont` associated with the context it was created from? I don't think so, but I could be wrong...

a-wi commented:
You are right. This way only graphics fonts created with `wxGraphicsContext::CreateFont()` are rescaled properly. Problem remains for those created with `wxGraphicsRenderer::CreateFont()` and assigned to the graphics context afterwards.
But we can intercept this font assignment in `wxGDIPlusPrintingContext::SetFont` and rescale font on-the-fly here. This seems to work in both cases. Something like this:
```diff
diff --git a/src/msw/graphics.cpp b/src/msw/graphics.cpp
--- a/src/msw/graphics.cpp
+++ b/src/msw/graphics.cpp
@@ -508,6 +508,8 @@ class wxGDIPlusPrintingContext : public wxGDIPlusContext
 public:
     wxGDIPlusPrintingContext( wxGraphicsRenderer* renderer, const wxDC& dc );
     virtual ~wxGDIPlusPrintingContext() { }
+
+    virtual void SetFont(const wxGraphicsFont& font) wxOVERRIDE;
 protected:
 };
 
@@ -2116,7 +2118,7 @@ void wxGDIPlusContext::GetTextExtent( const wxString &str, wxDouble *width, wxDo
 
     f->GetFamily(&ffamily) ;
 
-    REAL factorY = m_fontScaleRatio;
+    REAL factorY = 1.0;// m_fontScaleRatio;
 
     // Notice that we must use the real font style or the results would be
     // incorrect for italic/bold fonts.
@@ -2271,7 +2273,29 @@ wxGDIPlusPrintingContext::wxGDIPlusPrintingContext( wxGraphicsRenderer* renderer
 
     // We use this modifier when measuring fonts. It is needed because the
     // page scale is modified above.
-    m_fontScaleRatio = context->GetDpiY() / 72.0;
+    m_fontScaleRatio = context->GetDpiY() / 96.0;
+}
+
+void wxGDIPlusPrintingContext::SetFont(const wxGraphicsFont& font)
+{
+    Font* f = static_cast<wxGDIPlusFontData*>(font.GetRefData())->GetGDIPlusFont();
+    Brush* b = static_cast<wxGDIPlusFontData*>(font.GetRefData())->GetGDIPlusBrush();
+
+    // Retrieve parameters of native font
+    FontFamily family;
+    f->GetFamily(&family);
+    WCHAR name[LF_FACESIZE];
+    family.GetFamilyName(name);
+    REAL size = f->GetSize();
+    INT style = f->GetStyle();
+    Gdiplus::Color col;
+    static_cast<SolidBrush*>(b)->GetColor(&col);
+
+    // Create rescaled font
+    wxGraphicsFont newFont;
+    newFont.SetRefData(new wxGDIPlusFontData(GetRenderer(), name, size/m_fontScaleRatio, style,
+                                wxColour(col.GetR(), col.GetG(), col.GetB(), col.GetA())));
+    wxGDIPlusContext::SetFont(newFont);
 }
 
 //-----------------------------------------------------------------------------
```

a-wi commented:
With resized font it would be possible to get rid of pseudo-resizing done in `GetTextExtent what is a kind of advantage. Changing the unit is simpler and more elegant but the impact is global, not limited to the printing only. I really don't know which solution is better.

vadz committed:
Remove unnecessary wxGDIPlusPrintingContext dtor

No real changes.

vadz commented:
OK, looking at the 2 versions side by side, I think that the one based on your idea is more clear, so let's do it like this, thanks!

Please let me know if you have any objections, otherwise I'll merge this soon.

vadz committed:
Fix size of text drawing using wxGraphicsContext when printing

When using both wxEnhMetaFileDC (used for print preview) and
wxPrinterDC, a global scale factor was applied to the graphics context
in wxGDIPlusPrintingContext ctor and this changes size of the text
rendered using size specified in point units.

To prevent this from happening, stop applying this scale factor in the
base class GetTextExtent() and apply it instead when setting fonts in
the derived wxGDIPlusPrintingContext only. This ensures that the results
are consistent between GetTextExtent() and DrawText() and also keeps
the font scale hack entirely in wxGDIPlusPrintingContext without
affecting the base class.

It's still not totally clear why do we have to do this at all even
although we specify the font sizes in UnitPoint units which ought to be
DPI-independent -- but somehow are not. But at least the output is
correct now and the code is arguably more clear than before.

This fixes a regression with text size when printing or previewing
introduced by d5020362ffb99608eacd2492d621cad20b1075a7

Closes #18338.
